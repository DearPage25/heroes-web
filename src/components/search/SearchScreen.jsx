import { useLocation, useNavigate } from "react-router-dom";
import queryString from 'query-string'
import { useForm } from "../../hooks/useForm";
import { HeroCard } from '../hero/HeroCard'
import { getHeroesByName } from "../../selectors/getHeroesByName";
import { useMemo } from "react";


export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();
  
  const { q = '' } = queryString.parse(location.search);

  

  const [formValues, handleInputChange] = useForm({
    searchText: q,
  });

  const { searchText } = formValues
  const heroesFilter = useMemo(() => getHeroesByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${ searchText }`)
  }
  
  
  return (
    <>
      <h1>Search</h1>
        <hr />
        <div className="row">
          <div className="col-5">
            
            <form onSubmit={handleSearch}>
              <input
              type="text"
              placeholder="Search a Hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={handleInputChange}
              />

              <button
                className="btn btn-outline-primary mt-2"
                type="submit"
              >
                Search...
              </button>
            </form>
          </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {
            (q === '')
              ? <div className="alert alert-info">Search a Hero</div>
              : (heroesFilter.length === 0) 
                && <div className="alert alert-danger">There aren't result: { q }</div> 
          }
          {
            heroesFilter.map(hero => (
              <HeroCard
                key={hero.id}
                {...hero}
              />
            ))
          }
          </div>
        </div>
      </>
    )
  }