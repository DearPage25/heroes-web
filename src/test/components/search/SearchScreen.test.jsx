import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}))

describe('Tests in <SearchScreen />', () => {

    test('should be return the correctly component', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("Search a Hero");
    });
    test('should to show the batman', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find("input").prop("value")).toBe("batman");
        expect(wrapper).toMatchSnapshot();
    });
    test('should to show the error result', () => {
        const query = 'batman123'

        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${query}`]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find(".alert-danger").text().trim()).toBe(`There aren't result: ${query}`);
        expect(wrapper).toMatchSnapshot();
    });
    test('should to call navigate in the other screen', () => { 
        const value = 'batman';

        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search`]}>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: `${value}`
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault() { },
        });
        expect(mockNavigate).toHaveBeenCalledWith(`?q=${value}`)
    })


});