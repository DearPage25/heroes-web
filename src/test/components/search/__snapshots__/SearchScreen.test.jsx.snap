// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchScreen /> should be return the correctly component 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>,
]
`;

exports[`Tests in <SearchScreen /> should to show the batman 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                   
                  Batman
                   
                </h5>
                <p
                  className="card-text"
                >
                   
                  Bruce Wayne
                   
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  MÃ¡s...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Tests in <SearchScreen /> should to show the error result 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value="batman123"
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There aren't result: 
        batman123
      </div>
    </div>
  </div>,
]
`;
